### DISCUSSION TITLE
How to track user clicks in Ghidra?

### DISCUSSION THREAD

**Discussion by annetteshajan** [NationalSecurityAgency/ghidra#6023] [Q&A]
Hi 
I am trying to log user clicks in the Ghidra GUI.
Currently trying for the toolbar with the icons and shortcuts.
I understand we have to use `DockingToolActions` to access the toolbar, but how do we log every click on every Action item?
I have tried to iterate through it, I see a lot of users have added an addActionListener method to the DockingActionIf item, but I think it doesn't exist in the ghidradocs for me as a method under DockingActionIf.
Any advice would be appreciated!
Also let me know how to log clicks in general across the GUI if possible!

Thanks!

**Comment by dragonmacher** [comment]
There is no Ghidra API method for logging general user clicks. 

Java does provide a mechanism for tracking global events, which you can tap into to track items of interest.   We do this in a class called `DockingWindowsContextSensitiveHelpListener`, which has code that looks like this:

```java
   Toolkit toolkit = Toolkit.getDefaultToolkit();
        AWTEventListener listener = new AWTEventListener() {            
            public void eventDispatched( AWTEvent event ) {
                 Object itemUnderMouse = event.getSource();
                 System.out.println("mouse over item: " + itemUnderMouse);
            }
        }; 
        toolkit.addAWTEventListener( listener, AWTEvent.MOUSE_MOTION_EVENT_MASK );
```

In this example, the only event being tracked is mouse movement, not clicks, via the flag `AWTEvent.MOUSE_MOTION_EVENT_MASK `.

You should be able to do the same thing when the mouse is clicked.

**Reply by annetteshajan** [reply]
Sure I think this could work

But I only want to handle the clicks when the user clicks on the toolbar.
Is there any property that is altered when the user clicks on an action in DockingToolActions menu, so that I can just check if it is activated.

**Reply by dragonmacher** [reply]
There is no current property or event that we use to track such clicks. You should be able to write custom code, albeit brittle, that can detect the item that was clicked is a button belonging to the toolbar.

**Reply by annetteshajan** [reply]
Just want to know what is the current replacement for addActionListener?
